metrics/framework/Metrics.cpp

```diff
@@ -0,0 +1,212 @@
+    {
+        const char *epName = !isEmptyString(getInstanceProcName) ? getInstanceProcName : "getSinkInstance";
+        auto getInstanceProc = (getSinkInstance) GetSharedProcedure(libHandle, epName);
+        if (getInstanceProc != nullptr)
+        {
+            const char *name = isEmptyString(sinkName) ? type : sinkName;
+            pSink = getInstanceProc(name, pSettingsTree);
+        }
+    }
+    return pSink;
+}
+
+
+void MetricsReporter::init(IPropertyTree *pMetricsTree)
+{
+    initializeSinks(pMetricsTree->getElements("sinks"));
+}
+
+
+void MetricsReporter::addMetric(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(pMetric->getName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->getName(), pMetric});
+    }
+    // note that
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(metricName);
+    if (it != metrics.end())
+    {
+        metrics.erase(it);
+    }
+}
+
+
+void MetricsReporter::startCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->startCollection(this);
+    }
+}
+
+
+void MetricsReporter::stopCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->stopCollection();
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::getReportMetrics(const std::string &sinkName)
+{
+    std::vector<std::shared_ptr<IMetric>> reportMetrics;
+
+    auto it = sinks.find(sinkName);
+    if (it != sinks.end())
+    {
+        //
+        // This is where metric names are matched against patterns, however for now, just return all the metrics
+        std::unique_lock<std::mutex> lock(metricVectorMutex);   // no one else can mess with it for a bit
+        auto metricIt=metrics.begin();
+        while (metricIt != metrics.end())
+        {
+            auto pMetric = metricIt->second.lock();
+            if (pMetric)
+            {
+                reportMetrics.emplace_back(pMetric);
+                metricIt++;
+            }
+            else
+            {
+                metricIt = metrics.erase(metricIt);
+            }
+        }
+    }
+    // else throw? Detect that the sink Name is no longer valid?
+
+
+    return reportMetrics;  // All metrics
+}
+
+
+
+bool MetricsReporter::initializeSinks(IPropertyTreeIterator *pSinkIt)
+{
+    bool rc = true;
+    for (pSinkIt->first(); pSinkIt->isValid() && rc; pSinkIt->next())
+    {
+        SinkInfo *pSinkInfo;
+        IPropertyTree &sinkTree = pSinkIt->query();
+
+        StringBuffer cfgSinkType, cfgLibName, cfgSinkName, cfgProcName;
+        sinkTree.getProp("@type", cfgSinkType);  // this one is required
+        sinkTree.getProp("@libname", cfgLibName);
+        sinkTree.getProp("@name", cfgSinkName);
+        sinkTree.getProp("@instance_proc", cfgProcName);
```

Comment:
This structure isn't in the design document

Reply:
Added to design doc. Should we decide to not allow overriding of defaults for these values, we can remove them.